dataplane/memory_manager.cpp

```diff
@@ -0,0 +1,142 @@
+#include "memory_manager.h"
+
+using namespace dataplane;
+
+memory_pointer::memory_pointer(const char* name,
+                               const tSocketId socket_id,
+                               const size_t size,
+                               void* pointer,
+                               const std::function<void(void*)>& destructor) :
+        name(name),
+        socket_id(socket_id),
+        size(size),
+        pointer(pointer),
+        destructor(destructor)
+{
+}
+
+memory_pointer::~memory_pointer()
+{
+	YANET_LOG_INFO("yanet_free(name: '%s', socket: %u, size: %lu)\n",
+	               name.data(),
+	               socket_id,
+	               size);
+	destructor(pointer);
+}
+
+memory_manager::memory_manager() :
+        dataplane(nullptr)
+{
+}
+
+eResult memory_manager::init(cDataPlane* dataplane)
+{
+	this->dataplane = dataplane;
+	return eResult::success;
+}
+
+inline std::string to_hex(const void* pointer)
+{
+	char buffer[128];
+	snprintf(buffer, 128, "%p", pointer);
+	return buffer;
+}
+
+void memory_manager::report(nlohmann::json& json)
+{
+	std::lock_guard<std::mutex> guard(pointers_mutex);
+	for (const auto& [pointer, memory_pointer] : pointers)
+	{
+		nlohmann::json json_object;
+		json_object["pointer"] = to_hex(pointer);
+		json_object["name"] = memory_pointer.name;
+		json_object["socket_id"] = memory_pointer.socket_id;
+		json_object["size"] = memory_pointer.size;
+		json["memory_manager"].emplace_back(json_object);
+	}
+}
+
+eResult memory_manager::memory_manager_update(const common::idp::memory_manager_update::request& request)
+{
+	root_memory_group = request;
+	return eResult::success;
+}
+
+void* memory_manager::alloc(const char* name,
+                            const tSocketId socket_id,
+                            uint64_t size,
+                            const std::function<void(void*)>& destructor)
+{
+	if (!size)
+	{
+		YANET_LOG_ERROR("error allocation memory (name: '%s', socket: %u, size: %lu)\n",
+		                name,
+		                socket_id,
+		                size);
+		return nullptr;
+	}
+
+	size += 2 * RTE_CACHE_LINE_SIZE;
+
+	YANET_LOG_INFO("yanet_alloc(name: '%s', socket: %u, size: %lu)\n",
+	               name,
+	               socket_id,
+	               size);
+
+	void* pointer = rte_malloc_socket(nullptr,
```

Comment:
I do not like the rte_malloc here as it could be fragile, for instance if we are near the overall allocation limits.

Reply:
At minimum a complex investigation in rte_malloc behavior should be done before its application for dynamic allocation purposes.
Personally, I am unsure that allocation sequences like A16 A16, A40 F16 F16 A40 (allocate 16GB twice, then reallocate the first one from 16 to 40 and then the second one 16 to 40) would be handled well.

Reply:
We will not make things worse in this place